operator/iamkeycontroller/webhook_test.go

```diff
@@ -0,0 +1,380 @@
+							Namespace: "secret-namespace",
+						},
+					},
+					ForProvider: exoscalev1.IAMKeyParameters{
+						Services: exoscalev1.ServicesSpec{},
+					},
+				},
+			},
+			expectedError: "an IAMKey named \"iamkey-with-no-buckets\" should have at least 1 allowed bucket",
+		},
+		"GivenIAMKey_When2Buckets_ThenExpectNoError": {
+			iamKey: exoscalev1.IAMKey{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "iamkey-with-buckets",
+				},
+				Spec: exoscalev1.IAMKeySpec{
+					ResourceSpec: xpv1.ResourceSpec{
+						WriteConnectionSecretToReference: &xpv1.SecretReference{
+							Name:      "secret-name",
+							Namespace: "secret-namespace",
+						},
+					},
+					ForProvider: exoscalev1.IAMKeyParameters{
+						Services: exoscalev1.ServicesSpec{
+							SOS: exoscalev1.SOSSpec{
+								Buckets: []string{"bucket.1", "bucket.2"},
+							},
+						},
+					},
+				},
+			},
+			expectedError: "",
+		},
+		"GivenIAMKey_WhenWriteConnectionSecretToRefNameAndNamespaceExists_ThenExpectNoError": {
+			iamKey: exoscalev1.IAMKey{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "iamkey-with-buckets",
+				},
+				Spec: exoscalev1.IAMKeySpec{
+					ResourceSpec: xpv1.ResourceSpec{
+						WriteConnectionSecretToReference: &xpv1.SecretReference{
+							Name:      "secret-name",
+							Namespace: "secret-namespace",
+						},
+					},
+					ForProvider: exoscalev1.IAMKeyParameters{
+						Services: exoscalev1.ServicesSpec{
+							SOS: exoscalev1.SOSSpec{
+								Buckets: []string{"bucket.1", "bucket.2"},
+							},
+						},
+					},
+				},
+			},
+			expectedError: "",
+		},
+		"GivenIAMKey_WhenWriteConnectionSecretToRefMissing_ThenExpectError": {
+			iamKey: exoscalev1.IAMKey{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "iamkey-with-buckets",
+				},
+				Spec: exoscalev1.IAMKeySpec{
+					ResourceSpec: xpv1.ResourceSpec{
+						WriteConnectionSecretToReference: nil,
+					},
+					ForProvider: exoscalev1.IAMKeyParameters{
+						Services: exoscalev1.ServicesSpec{
+							SOS: exoscalev1.SOSSpec{
+								Buckets: []string{"bucket.1", "bucket.2"},
+							},
+						},
+					},
+				},
+			},
+			expectedError: "an IAMKey named \"iamkey-with-buckets\" should have a connection secret reference",
+		},
+		"GivenIAMKey_WhenWriteConnectionSecretToRefNameMissing_ThenExpectError": {
+			iamKey: exoscalev1.IAMKey{
+				TypeMeta: metav1.TypeMeta{},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "iamkey-with-buckets",
+				},
+				Spec: exoscalev1.IAMKeySpec{
+					ResourceSpec: xpv1.ResourceSpec{
+						WriteConnectionSecretToReference: &xpv1.SecretReference{
+							Name:      "",
+							Namespace: "secret-namespace",
+						},
+					},
+					ForProvider: exoscalev1.IAMKeyParameters{
+						Services: exoscalev1.ServicesSpec{
+							SOS: exoscalev1.SOSSpec{
+								Buckets: []string{"bucket.1", "bucket.2"},
+							},
+						},
+					},
```

Comment:
I would give to the unit test the correct object minus the part which I want to test. In this case the buckets should be present otherwise it may fail for a different reason than the test intended in the first place.

Reply:
I see